### ISSUE TITLE
[strict-boolean-expressions] Possible new option allowing nullable Array.length

### ISSUE THREAD

**Issue by jakubmazanec** [typescript-eslint/typescript-eslint#3323]
I have tried searching if this was discussed before but was unable to find anything, so here it goes.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": "error"
  }
}
```

```TS
// 1)

if (foo?.length) { // strict-boolean-expressions error
  // I only want to run this when foo exists and is not empty
}

if (foo?.length > 0) { // TypeScript error
  // I only want to run this when foo exists and is not empty
}

if (foo && foo.length > 0) { // ok, but longer code
  // I only want to run this when foo exists and is not empty
}

// 2)

if (!foo?.length) { // strict-boolean-expressions error
  // I only want to run this when foo doesn't exists or is empty
}

if (foo === undefined || foo.length === 0) { // ok, but longer code
  // I only want to run this when foo doesn't exists or is empty
}
```

Correct me if I'm wrong, but I think it would be useful to have additional option `allowNullableArrayLength` (or something like that) that allows nullable number, if it's `length` property of an array, because in that case we don't care about differentiating between undefined/null and 0.

**Comment by bradzacher** [comment]
> property of an array, because in that case we don't care about differentiating between undefined/null and 0

What's the difference between an nullish array and any nullish object with a property that is a number - say a `Map` or a `Set`? What about in comparison to a nullish number typed variable?

```ts
function arr(arg: string[] | null) {
  if (arg?.length) { ... }
}

function set(arg: Set<string> | null) {
  if (arg?.size) { ... }
}

function map(arg: Map<string> | null) {
  if (arg?.size) { ... }
}

class Custom { num: number }
function map(arg: Custom | null) {
  if (arg?.num) { ... }
}

function num(arg: number | null) {
  if (arg) { ... }
}
```

The answer is that there isn't one - they're all the same - they're all a case of an expression that has multiple falsey cases.

The question is - do you as an engineer care about strictly handling nullish numbers?

If the answer is yes - you want to be strict, then the rule is working fine as is - you must strictly handle your nullish state before your number state - `arg != null && arg.length`

If the answer is no - you don't want to be strict, then the rule is still working fine as is, and [you can reconfigure the rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md#allownullablenumber) to suit your coding style.

**Comment by jakubmazanec** [comment]
@bradzacher I really don't understand your reasoning. I'm not talking about "any nullish object with a property that is a number", I'm talking about an array with a property called 'length' that is a number. And I do care about strictly handling nullish numbers - just not in this case.

**Comment by bradzacher** [comment]
I understand exactly what you want here.

My question is this:
"What is the difference between the array's length property, and a property of type number on any other object?"

or put another way:
"Why should the rule special case `Array`'s `length` property, but not the `Map`/`Set`'s `size` property?"

I suspect that your answer is along the lines of
"In my codebase I don't care about differentiating between an empty array and a null array"
Which is a valid reason for the request - for sure.

But my response to that is that it is not a good reason to add an option to the rule.

**Comment by jakubmazanec** [comment]
Honestly, why not also .size? The reasoning's the same. I mean, the rule is really useful, but still not granular enough. I remember the first version of the rule; the new version is much better, why not continue the trend? (Anyway, thanks for finding time to answer my questions!)

**Comment by rajzik** [comment]
@bradzacher 
Due to the nature of javascript `length` and `size` property, they can be considered special. 

In most cases when you have falsy value of these properties you want the same behavior, I believe even strings. 

```ts
const maybeArray: unknown[] | undefined;

if (maybeArray?.length) {
   maybeArray.map(...);  
}

const maybeMap: Map<string, unknown> | undefined;

if (maybeMap?.size) {
  maybeMap.forEach(...);
}


const maybeSet: Set<unknown> | undefined;

if (maybeSet?.size) {
  maybeSet.forEach(...);
}

const maybeString: string | undefined;

if (maybeString?.length) {
  maybeString.trim();
}
```

All these examples above seems valid and can be useful in most (of our) cases.
